### PR TITLE
executor: make the aggregation function `distinct` consider collations

### LINKED ISSUE #16668: collation: the aggregate function `distinct` doesn't consider collations
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?

```
CREATE TABLE `t` ( `a` int(11) NOT NULL,`b` varchar(5) CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci DEFAULT NULL);
insert into t values (1, 'a'), (2, 'À'), (3, 'á'), (4, 'à'), (5, 'b'), (6, 'c'), (7, ' ');
select count(distinct(b)) from t;
```

### 2. What did you expect to see?
```
mysql> select count(distinct(b)) from t;
+--------------------+
| count(distinct(b)) |
+--------------------+
|                  4 |
+--------------------+
```

### 3. What did you see instead?
```
mysql> select count(distinct(b)) from t;
+--------------------+
| count(distinct(b)) |
+--------------------+
|                  7 |
+--------------------+

```

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #16668 

Problem Summary: the aggregation function `distinct` doesn't consider collations

### What is changed and how it works?
Check collation when comparing strings in aggregation function `distinct`.

### Check List 

Tests 

- Unit test

### Release note 
- make the aggregation function `distinct` consider collations

### CODE REVIEW

File: executor/aggfuncs/func_count_distinct.go [hunk 1]
```diff
@@ -242,6 +243,8 @@ func (e *countOriginalWithDistinct4String) AppendFinalResult2Chunk(sctx sessionc
 
 func (e *countOriginalWithDistinct4String) UpdatePartialResult(sctx sessionctx.Context, rowsInGroup []chunk.Row, pr PartialResult) error {
 	p := (*partialResult4CountDistinctString)(pr)
+	_, coll := e.args[0].CharsetAndCollation(sctx)
```
Comment (wjhuang2016) [inline executor/aggfuncs/func_count_distinct.go:246]: We should use the result type of the first argument directly.

Comment (qw4990) [reply executor/aggfuncs/func_count_distinct.go:246]: You mean that we use `e.args[0].RetType.Collation` here?

Comment (wjhuang2016) [reply executor/aggfuncs/func_count_distinct.go:246]: Yeah

Comment (qw4990) [reply executor/aggfuncs/func_count_distinct.go]: updated, PTAL.

**Comment by qw4990** [comment]
> `countOriginalWithDistinct`, `groupConcatDistinct` and `groupConcatDistinctOrder` may be necessary to update.

I will create a new issue for these functions and fix them later.

**Pull request merged**